lightning/src/ln/channel_state.rs

```diff
@@ -549,128 +546,45 @@ impl ChannelDetails {
-		let user_channel_id_low: u64 = user_channel_id_low.0.unwrap();
-		let user_channel_id = user_channel_id_low as u128
-			+ ((user_channel_id_high_opt.unwrap_or(0 as u64) as u128) << 64);
-
-		let _balance_msat: Option<u64> = _balance_msat;
-
-		Ok(Self {
-			inbound_scid_alias,
-			channel_id: channel_id.0.unwrap(),
-			channel_type,
-			counterparty: counterparty.0.unwrap(),
-			outbound_scid_alias,
-			funding_txo,
-			config,
-			short_channel_id,
-			channel_value_satoshis: channel_value_satoshis.0.unwrap(),
-			unspendable_punishment_reserve,
-			user_channel_id,
-			outbound_capacity_msat: outbound_capacity_msat.0.unwrap(),
-			next_outbound_htlc_limit_msat: next_outbound_htlc_limit_msat.0.unwrap(),
-			next_outbound_htlc_minimum_msat: next_outbound_htlc_minimum_msat.0.unwrap(),
-			inbound_capacity_msat: inbound_capacity_msat.0.unwrap(),
-			confirmations_required,
-			confirmations,
-			force_close_spend_delay,
-			is_outbound: is_outbound.0.unwrap(),
-			is_channel_ready: is_channel_ready.0.unwrap(),
-			is_usable: is_usable.0.unwrap(),
-			is_announced: is_announced.0.unwrap(),
-			inbound_htlc_minimum_msat,
-			inbound_htlc_maximum_msat,
-			feerate_sat_per_1000_weight,
-			channel_shutdown_state,
-			pending_inbound_htlcs: pending_inbound_htlcs.unwrap_or(Vec::new()),
-			pending_outbound_htlcs: pending_outbound_htlcs.unwrap_or(Vec::new()),
-		})
-	}
-}
+impl_writeable_tlv_based!(ChannelDetails, {
+	(1, inbound_scid_alias, option),
+	(2, channel_id, required),
+	(3, channel_type, option),
+	(4, counterparty, required),
+	(5, outbound_scid_alias, option),
+	(6, funding_txo, option),
+	(7, config, option),
+	(8, short_channel_id, option),
+	(9, confirmations, option),
+	(10, channel_value_satoshis, required),
+	(12, unspendable_punishment_reserve, option),
+	// Note that _user_channel_id_low is used below, but rustc warns anyway
+	(14, _user_channel_id_low, (legacy, u64,
+		|us: &ChannelDetails| Some(us.user_channel_id as u64))),
+	(16, _balance_msat, (legacy, u64, |us: &ChannelDetails| Some(us.next_outbound_htlc_limit_msat))),
+	(18, outbound_capacity_msat, required),
+	(19, next_outbound_htlc_limit_msat, (default_value, outbound_capacity_msat)),
+	(20, inbound_capacity_msat, required),
+	(21, next_outbound_htlc_minimum_msat, (default_value, 0)),
+	(22, confirmations_required, option),
+	(24, force_close_spend_delay, option),
+	(26, is_outbound, required),
+	(28, is_channel_ready, required),
+	(30, is_usable, required),
+	(32, is_announced, required),
+	(33, inbound_htlc_minimum_msat, option),
+	(35, inbound_htlc_maximum_msat, option),
+	// Note that _user_channel_id_high is used below, but rustc warns anyway
+	(37, _user_channel_id_high, (legacy, u64,
+		|us: &ChannelDetails| Some((us.user_channel_id >> 64) as u64))),
+	(39, feerate_sat_per_1000_weight, option),
+	(41, channel_shutdown_state, option),
+	(43, pending_inbound_htlcs, optional_vec),
+	(45, pending_outbound_htlcs, optional_vec),
+	(_unused, user_channel_id, (static_value,
```

Comment:
Could we use `_` instead of `_unused`? The latter leaks implementation details of the macro and may be confusing to someone reading this, like mistakenly thinking `user_channel_id` is unused.